### ISSUE TITLE
Encouter issue when trying to reproduce "Learning high-accuracy error decoding for quantum processors"

### ISSUE THREAD

**Issue by Chenning-Tao** [Dominik5431/QecTransformer#1]
Hi there, I'm also trying to build a transformer-based decoder for QEC based on Google's AlphaQubit paper. However, I have encountered some problems. I'm wondering if you can help me with it.
1. Using the DEM data, the network will predict the same result no matter the input, and I find the gradient of the embedding layer would be very large. I have tried your code, and you have similar issues. The loss wouldn't go down.
2. Using the Pauli+ data, the network's gradient will either be 0 or nan, making it impossible to train.

**Comment by Dominik5431** [comment]
Hi,
sure!
1. I assume this happens especially for larger distances? I'm still working on this, but the issue is, that in your trainings data each syndrome appears to rarely. The decoder rather ignores the encoder input aka the syndrome and produces its output ignoring the input. I can only guess but I think, the Google people encountered the same issue but starting from distance 11 or 13 (they write it at some point that training for d=11 got severely more difficult). My guess is, they managed to go up to 11 due to the immense amount of data they're training on. With that they ensure to have each syndrome sufficiently often in their training data. But obviously, this doesn't scale arbitrarily high. Citing from their outlook they also write, that some things about training or structure need to be changed to scale to higher distances.
2. I'm not exactly sure, where this is coming from, but you might try gradient clipping. Usually, nan is caused by exploding gradients.

**Comment by Chenning-Tao** [comment]
I found some differences between your implementation and the one described in Google's paper, and my implementation is the decoder module. 
1. Your default output is using the Transformer's decoder module. The method described in Google's paper uses a readout network, which may be similar to your code's 'conv' part. But when selecting 'conv,' the output shape didn't match.
2. The implementation in Google's paper encodes the syndrome one step at a time, but you seem to take the whole syndrome at once.
I wonder what the considerations behind those changes are. 
I followed the pseudo code provided in Google's paper for my implementation, but I encountered this problem even when the code distance was 3. The network will simply predict 0 no matter what the input is. When I print out the gradient, I find the gradient seems to grow exponentially. There is one part of the implementation that is mentioned in Google's paper that bothers me. For the input, they basically expanded one dimension 0/1 into a 320-dimensional input vector using linear layers, which is very odd.

Regarding the lack of data you mentioned, I have used the Pauli+ dataset provided by Google, which contains billions of training examples, but the gradient goes to nan. I figure it may have something to do with the round=25.
